site/en/docs/android/custom-tabs/headers/index.md

```diff
@@ -0,0 +1,218 @@
+
+```java
+// Set up a connection that warms up and validates a session.
+CustomTabsServiceConnection connection = new CustomTabsServiceConnection() {
+    @Override
+    public void onCustomTabsServiceConnected(@NonNull ComponentName name, 
+        @NonNull CustomTabsClient client) {
+        // Create session after service connected.
+        mSession = client.newSession(callback);
+        client.warmup(0);
+        // Validate the session as the same origin to allow cross origin headers.
+        mSession.validateRelationship(CustomTabsService.RELATION_USE_AS_ORIGIN, 
+            Uri.parse(url), null);
+    }
+    @Override
+    public void onServiceDisconnected(ComponentName componentName) { }
+};
+```
+
+### Set up a Callback that Launches the Intent after Validation
+
+The `CustomTabsCallback` was passed into the session. We set up its
+`onRelationshipValidationResult()` to launch the previously created `CustomTabsIntent`
+once the origin verification succeeds.
+
+```java
+// Set up a callback that launches the intent after session validated.
+CustomTabsCallback callback = new CustomTabsCallback() {
+    @Override
+    public void onRelationshipValidationResult(int relation, @NonNull Uri requestedOrigin, 
+        boolean result, @Nullable Bundle extras) {
+        // Launch custom tabs intent after session was validated as the same origin.
+        CustomTabsIntent intent = constructExtraHeadersIntent(mSession);
+        intent.launchUrl(MainActivity.this, Uri.parse(url));
+    }
+};
+```
+
+### Bind the custom tabs service connection
+
+Binding the service launches the service and the connection's `onCustomTabsServiceConnected()`
+will be called eventually. Don't forget to unbind the service appropriately. Binding and unbinding
+is commonly done in the `onStart()` and `onStop()` activity lifecycle methods.
+
+```java
+// Bind the custom tabs service connection.
+// Call this in onStart()
+CustomTabsClient.bindCustomTabsService(this,
+    CustomTabsClient.getPackageName(MainActivity.this, null), connection);
+
+// …
+// Unbind the custom tabs service.
+// Call this in onStop().
+unbindService(connection);
+```
+
+### Demo application code
+
+You can find more details about Custom Tabs Service [here][12]. See the
+[android-browser-helper][13] GitHub repository for a working example app.
+
+## Summary
+This guide demonstated how to add arbitrary headers to custom tabs CORS requests.
+Whitelisted headers can be attached to every custom tabs CORS request. Non-whitelisted headers are
+generally considered unsafe in CORS requests and chrome filters them by default. Attaching them is
+allowed only for clients and servers of the same origin, verified by a digital asset link.
+
+[1]: https://developer.android.com/reference/android/provider/Browser#EXTRA_HEADERS
+[2]: https://docs.google.com/document/d/1sN6Y31giDbdGj6R4p3QJ20gD1ggBIosae7yrsSZ3Ins/edit#heading=h.ey2vxjsxytw6
+[3]: https://developer.chrome.com/multidevice/android/customtabs
```

Comment:
And update this URL and the ones below to start with "/".

Reply:
done.